From 7f9f3d03612fd72735b96e212993ce0531af976d Mon Sep 17 00:00:00 2001
From: Florian Westphal <fwestpha@redhat.com>
Date: Tue, 2 Sep 2014 10:14:34 -0400
Subject: [net] gre: Call skb_checksum_simple_validate

Message-id: <1409652920-431-12-git-send-email-fwestpha@redhat.com>
Patchwork-id: 90174
O-Subject: [PATCH 11/57] [RHEL7.1 net] gre: Call skb_checksum_simple_validate
Bugzilla: 1131999
RH-Acked-by: Jiri Pirko <jpirko@redhat.com>
RH-Acked-by: Vlad Yasevich <vyasevic@redhat.com>
RH-Acked-by: Nikolay Aleksandrov <nikolay@redhat.com>
RH-Acked-by: Ivan Vecera <ivecera@redhat.com>
RH-Acked-by: Jiri Benc <jbenc@redhat.com>

BZ: 1131999

commit b1036c6a470ccf5f18490a7ce4c99422d3bf77c4
Author: Tom Herbert <therbert@google.com>
Date:   Wed May 7 16:51:46 2014 -0700

    gre: Call skb_checksum_simple_validate

    Use skb_checksum_simple_validate to verify checksum.

    Signed-off-by: Tom Herbert <therbert@google.com>
    Signed-off-by: David S. Miller <davem@davemloft.net>

Signed-off-by: Jiri Benc <jbenc@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/net/ipv4/gre_demux.c b/net/ipv4/gre_demux.c
index 1863422..8bf3b4d 100644
--- a/net/ipv4/gre_demux.c
+++ b/net/ipv4/gre_demux.c
@@ -93,28 +93,6 @@ void gre_build_header(struct sk_buff *skb, const struct tnl_ptk_info *tpi,
 }
 EXPORT_SYMBOL_GPL(gre_build_header);
 
-static __sum16 check_checksum(struct sk_buff *skb)
-{
- __sum16 csum = 0;
-
- switch (skb->ip_summed) {
- case CHECKSUM_COMPLETE:
-  csum = csum_fold(skb->csum);
-
-  if (!csum)
-   break;
-  /* Fall through. */
-
- case CHECKSUM_NONE:
-  skb->csum = 0;
-  csum = __skb_checksum_complete(skb);
-  skb->ip_summed = CHECKSUM_COMPLETE;
-  break;
- }
-
- return csum;
-}
-
 static int parse_gre_header(struct sk_buff *skb, struct tnl_ptk_info *tpi,
        bool *csum_err)
 {
@@ -141,7 +119,7 @@ static int parse_gre_header(struct sk_buff *skb, struct tnl_ptk_info *tpi,
 
  options = (__be32 *)(greh + 1);
  if (greh->flags & GRE_CSUM) {
-  if (check_checksum(skb)) {
+  if (skb_checksum_simple_validate(skb)) {
    *csum_err = true;
    return -EINVAL;
   }
-- 
1.7.1